File: internal/db/model/check.go
package model

import (
	"time"

	"github.com/ZeusWPI/events/internal/db/sqlc"
)

type CheckStatus string

const (
	CheckDone     CheckStatus = "done"      // Check is furfilled
	CheckDoneLate CheckStatus = "done_late" // Check is furfilled too late but without too many consequences
	CheckTODO     CheckStatus = "todo"      // Still needs to be done
	CheckTODOLate CheckStatus = "todo_late" // Still needs to be done but already too late
	CheckWarning  CheckStatus = "warning"   // Something is up with the check
)

type CheckType string

const (
	CheckManual    CheckType = "manual"
	CheckAutomatic CheckType = "automatic"
)

type Check struct {
	// Check event fields
	ID         int // ID of the check event
	EventID    int
	Status     CheckStatus
	Message    string
	Mattermost bool
	UpdatedAt  time.Time

	// Check fields
	UID         string // Identifier of the check
	Description string
	Deadline    time.Duration
	Active      bool
	Type        CheckType
	CreatorID   int
}

func CheckModel(check sqlc.Check, checkEvent sqlc.CheckEvent) *Check {
	message := ""
	if checkEvent.Message.Valid {
		message = checkEvent.Message.String
	}

	deadline := time.Duration(0)
	if check.Deadline.Valid {
		deadline = time.Duration(check.Deadline.Int64)
	}

	creatorID := 0
	if check.CreatorID.Valid {
		creatorID = int(check.CreatorID.Int32)
	}

	return &Check{
		ID:         int(checkEvent.ID),
		EventID:    int(checkEvent.EventID),
		Status:     CheckStatus(checkEvent.Status),
		Message:    message,
		Mattermost: checkEvent.Mattermost,
		UpdatedAt:  checkEvent.UpdatedAt.Time,

		UID:         check.Uid,
		Description: check.Description,
		Deadline:    deadline,
		Active:      check.Active,
		Type:        CheckType(check.Type),
		CreatorID:   creatorID,
	}
}

File: ui/src/lib/types/check.ts
import { SquareCheckIcon, SquareIcon, SquareXIcon, TriangleAlertIcon } from "lucide-react";
import { createElement } from "react";
import { API } from "./api";
import { Base } from "./general";

export enum CheckStatus {
  Done = "done",
  DoneLate = "done_late",
  Todo = "todo",
  TodoLate = "todo_late",
  Warning = "warning",
}

export const checkStatusToIcon: Record<CheckStatus, React.ReactNode> = {
  [CheckStatus.Done]: createElement(SquareCheckIcon, { className: 'text-green-500' }),
  [CheckStatus.DoneLate]: createElement(SquareCheckIcon, { className: 'text-orange-500' }),
  [CheckStatus.Todo]: createElement(SquareIcon),
  [CheckStatus.TodoLate]: createElement(SquareXIcon, { className: 'text-red-500' }),
  [CheckStatus.Warning]: createElement(TriangleAlertIcon, { className: 'text-orange-500' }),
}

export const checkStatusToText: Record<CheckStatus, string> = {
  [CheckStatus.Done]: "Done",
  [CheckStatus.DoneLate]: "Done after deadline",
  [CheckStatus.Todo]: "Todo",
  [CheckStatus.TodoLate]: "Deadline passed",
  [CheckStatus.Warning]: "Warning",
}

export enum CheckType {
  Manual = "manual",
  Automatic = "automatic",
}

export interface Check extends Base {
  eventId: number;
  status: CheckStatus;
  message?: string;
  description: string;
  deadline?: number;
  type: CheckType;
  creator_id?: number;
}

export function convertCheckToModel(check: API.Check): Check {
  return {
    id: check.id,
    eventId: check.event_id,
    status: check.status as CheckStatus,
    message: check.message,
    description: check.description,
    deadline: check.deadline,
    type: check.type as CheckType,
    creator_id: check.creator_id,
  };
}

